### ISSUE TITLE
Can not reprodece csresnext50-panet-spp-original-optimal Results.

### ISSUE THREAD

**Issue by tianfengyijiu** [AlexeyAB/darknet#5060]
Hi,
I use csresnext50-panet-spp-original-optimal to train coco2017:
![image](https://user-images.githubusercontent.com/5301100/76928649-2d1c0300-691d-11ea-9b94-da6051a76be7.png)

I use 8 2080ti GPUs.But can not get 65.4mAP50.Test mAP with 608 image get :
![image](https://user-images.githubusercontent.com/5301100/76928792-7ec48d80-691d-11ea-9961-14bb624148eb.png)

What wrong happens

**Comment by WongKinYiu** [comment]
In your cfg: `width=416`, `height=416`.

**Comment by tianfengyijiu** [comment]
@WongKinYiu Yes, I modify the width and height from 608 to 416, because if 608 will lead to CUDA out of memory with multi-scale.
command is :
![image](https://user-images.githubusercontent.com/5301100/76952681-6ae25100-6948-11ea-85b2-6621c1421694.png)
Thanks a lot!

**Comment by WongKinYiu** [comment]
The anchor size of csresnext50-panet-spp-original-optimal is optimized for 512x512.
If you want to train with 416x416, the anchors should be 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326.
And the AP50 will decrease about 1%.

Also, darknet53.conv.74 is the pretrained model for YOLOv3.
For csresnext50-panet-spp-original-optimal, you should use csresnext50.conv.80 as pretrained model.

**Comment by AlexeyAB** [comment]
@tianfengyijiu 

Download
* cfg https://raw.githubusercontent.com/WongKinYiu/CrossStagePartialNetworks/master/cfg/csresnext50.cfg
* weights https://drive.google.com/open?id=1IxKu5lAYCo4UpzAl5pOBIDAvC843vjn1

Then do:
`./darknet partial csresnext50.cfg csresnext50.weights csresnext50.conv.80 80`

And train by using `csresnext50.conv.80`

`./darknet detector train cfg/coco.data cfg/csresnext50-panet-spp-original-optimal.cfg csresnext50.conv.80 -map -gpus 0,1,2,3,4,5,6,7`

Use these settings in https://raw.githubusercontent.com/AlexeyAB/darknet/master/cfg/csresnext50-panet-spp-original-optimal.cfg
```
batch=64
subdivisions=16
width=512
height=512
```


After training set 
```
width=512
height=512
```

**Comment by tianfengyijiu** [comment]
@AlexeyAB Thanks a lot, but I can not train with 512 because CUDA out of memory, I will try with 416 after modifying anchors.
Which will be better trained in 416 with multi-scale or train in 512 without multi-scale?

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

You said to get 112 layers for making a pre-trained in this issue [csresnext50-panet-spp-original-optimal.conv.112](https://github.com/AlexeyAB/darknet/issues/4974#issuecomment-596270598) but why are saying to use 80 layers? 
Which pre-trained is correct because I get avg -nan after about 3k or 5k iter?

Thanks

**Comment by WongKinYiu** [comment]
because in this issue, the problem is "reproduce csresnext50-panet-app-original-optimal results".
so the answer is using imagenet pretrained model, which is csresnext50.conv.80.

however, csresnext50-panet-spp-original-optimal.conv.112 is already trained on coco dataset.
it can be used as pretrained model for training custom dataset.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi

I have ~320 classes and my classes are traffic signs. I'm going to use a model in real-time. Sometimes it detected wrong especially on speed limit signs and some signs that are similar in shape. Could you please recommend me for this case?

**Comment by WongKinYiu** [comment]
could those traffic signs be detected correctly using a model not in real-time?

**Comment by zpmmehrdad** [comment]
I tested csresnext50-panet-spp-original-optimal. The result was good but when I test on a new videos that some signs are not in my classes the model is detecting the sign 100% if it is wrong. I want to reduce the false. And another problem that I have on training mode sometimes after 3 or 5 I get avg nan.

Thanks

**Comment by WongKinYiu** [comment]
this is the shortage of classification loss, for example, if there are only dogs in your training images, then if you input a test image which contains a wolf, the wolf is easier to be identified as a dog.

so in your case, every traffic signs are easier to be identified as 1 of ~320 classes of traffic signs.
the most simple solution is put some traffic signs which are not included in ~320 classes as negative examples in your dataset.
a possible solution is setting different threshold of every classes.
the best solution is using loss function of metric learning or adversarial learning, it can find the bound of each classes in feature space.

for nan issue, the main problem is usually caused by bad setting. for example in your mentioned issue https://github.com/AlexeyAB/darknet/issues/5060#issuecomment-601058977, it is because the anchor size and grid size are mismatch. so just set the reasonable hyper-parameter and train your model.

**Comment by zpmmehrdad** [comment]
@WongKinYiu Hi,

Thanks for the reply, Is csresnext50-panet-spp-original-optimal good for this case and does it need to change the parameters, if yes which parameters? How can I change the loss function of metric learning or adversarial learning?

Thanks

**Comment by WongKinYiu** [comment]
hyper-parameters are depend on both of your model and dataset, i can not answer the question if i do not have information of your dataset.

modify loss function in detector.c, and also modify data loader to fit your new loss function.

**Comment by zpmmehrdad** [comment]
> I tested csresnext50-panet-spp-original-optimal. The result was good but when I test on a new videos that some signs are not in my classes the model is detecting the sign 100% if it is wrong. I want to reduce the false. And another problem that I have on training mode sometimes after 3 or 5 I get avg nan.
> 
> Thanks

@AlexeyAB Hi,

Could you please help me about this issue? I want to make the classification stronger and decrease the false. How can I do that?

**Comment by AlexeyAB** [comment]
@zpmmehrdad Hi, Create separate issue, add examples and description.

**Comment by jfkkf123** [comment]
> the best solution is using loss function of metric learning or adversarial learning, it can find the bound of each classes in feature space.

@WongKinYiu Hi，Could you recommend some papers about loss function of metric learning or adversarial learning ?Does it mean GAN?
thanks

**Comment by WongKinYiu** [comment]
for example, triplet loss.

**Comment by jfkkf123** [comment]
> for example, triplet loss.

thank you

**Comment by tianfengyijiu** [comment]
@AlexeyAB Hi, I do not know how to get mAP in COCO test-dev, Can you tell me the results of csresnext50-panet-spp-original-optimal in COCO val-set?

**Comment by WongKinYiu** [comment]
**AP50** of csresnext50-panet-spp-original-optimal on COCO val5k set is **64.1%**.

set `valid =` in your .data file.

for val5k set, `valid = 5k.txt`.
use `./darknet detector map coco.data coco.cfg coco.weights`.
then you can get AP50 of val5k set.

for test-dev set, `valid = testdev2017.txt` (evaluation sever of 2014 and 2015 are closed).
use `./darknet detector valid coco.data coco.cfg coco.weights`.
then you will get `results.json`, compress it to `result.zip` and submit to codalab competition to get result of test-dev set.
